### ISSUE TITLE
CycleGAN model showing little higher loss if the training is resumed after interruption 

### ISSUE THREAD

**Issue by nanditab35** [junyanz/pytorch-CycleGAN-and-pix2pix#1380]
CycleGAN model showing little higher loss if the training is resumed after interruption. My training was interrupted after saving checkpoint for epoch 177, then I resumed it to start running from 178 up to 200. But the 200 epoch model checkpoint result got less evaluation score while inference than the 177 epoch model checkpoint. 
Please let me know if I am missing anything. Thanks in Advance!
Given below is the log:

(epoch: 176, iters: 800, time: 0.311, data: 0.000) D_A: 0.091 G_A: 0.563 cycle_A: 0.605 idt_A: 0.203 D_B: 0.104 G_B: 0.562 cycle_B: 0.664 idt_B: 0.171 
(epoch: 176, iters: 900, time: 0.311, data: 0.000) D_A: 0.124 G_A: 1.365 cycle_A: 0.543 idt_A: 0.156 D_B: 0.184 G_B: 0.340 cycle_B: 0.519 idt_B: 0.157 
(epoch: 176, iters: 1000, time: 0.311, data: 0.000) D_A: 0.100 G_A: 0.453 cycle_A: 0.514 idt_A: 0.200 D_B: 0.122 G_B: 0.629 cycle_B: 0.575 idt_B: 0.213 
(epoch: 176, iters: 1100, time: 0.310, data: 0.000) D_A: 0.074 G_A: 0.849 cycle_A: 0.573 idt_A: 0.137 D_B: 0.262 G_B: 0.469 cycle_B: 0.417 idt_B: 0.183 
(epoch: 176, iters: 1200, time: 0.311, data: 0.000) D_A: 0.113 G_A: 0.628 cycle_A: 0.612 idt_A: 0.193 D_B: 0.142 G_B: 0.538 cycle_B: 0.361 idt_B: 0.240 
(epoch: 176, iters: 1300, time: 0.311, data: 0.000) D_A: 0.118 G_A: 0.817 cycle_A: 0.857 idt_A: 0.117 D_B: 0.122 G_B: 0.304 cycle_B: 0.432 idt_B: 0.252 
(epoch: 176, iters: 1400, time: 0.311, data: 0.000) D_A: 0.084 G_A: 0.711 cycle_A: 0.683 idt_A: 0.211 D_B: 0.136 G_B: 0.499 cycle_B: 0.634 idt_B: 0.213 
(epoch: 176, iters: 1500, time: 0.311, data: 0.000) D_A: 0.072 G_A: 0.844 cycle_A: 0.758 idt_A: 0.272 D_B: 0.235 G_B: 0.369 cycle_B: 0.808 idt_B: 0.244 
(epoch: 176, iters: 1600, time: 0.310, data: 0.000) D_A: 0.076 G_A: 0.995 cycle_A: 0.537 idt_A: 0.202 D_B: 0.177 G_B: 0.571 cycle_B: 0.592 idt_B: 0.174 
(epoch: 176, iters: 1700, time: 0.308, data: 0.000) D_A: 0.062 G_A: 0.843 cycle_A: 0.678 idt_A: 0.231 D_B: 0.210 G_B: 0.654 cycle_B: 0.522 idt_B: 0.167 
(epoch: 176, iters: 1800, time: 0.311, data: 0.000) D_A: 0.136 G_A: 0.436 cycle_A: 0.921 idt_A: 0.169 D_B: 0.161 G_B: 0.298 cycle_B: 0.516 idt_B: 0.240 
(epoch: 176, iters: 1900, time: 0.312, data: 0.000) D_A: 0.049 G_A: 0.943 cycle_A: 0.683 idt_A: 0.137 D_B: 0.052 G_B: 0.533 cycle_B: 0.457 idt_B: 0.210 
(epoch: 176, iters: 2000, time: 0.311, data: 0.000) D_A: 0.091 G_A: 0.486 cycle_A: 0.408 idt_A: 0.179 D_B: 0.151 G_B: 0.579 cycle_B: 0.533 idt_B: 0.131 
(epoch: 177, iters: 100, time: 0.312, data: 0.000) D_A: 0.107 G_A: 0.710 cycle_A: 0.608 idt_A: 0.149 D_B: 0.170 G_B: 0.312 cycle_B: 0.459 idt_B: 0.189 
(epoch: 177, iters: 200, time: 0.311, data: 0.000) D_A: 0.224 G_A: 0.645 cycle_A: 0.736 idt_A: 0.192 D_B: 0.215 G_B: 0.418 cycle_B: 0.584 idt_B: 0.279 
(epoch: 177, iters: 300, time: 0.320, data: 0.000) D_A: 0.072 G_A: 1.123 cycle_A: 0.525 idt_A: 0.150 D_B: 0.159 G_B: 0.363 cycle_B: 0.467 idt_B: 0.144 
(epoch: 177, iters: 400, time: 0.312, data: 0.000) D_A: 0.052 G_A: 0.556 cycle_A: 0.561 idt_A: 0.223 D_B: 0.177 G_B: 0.490 cycle_B: 0.560 idt_B: 0.165 
(epoch: 177, iters: 500, time: 0.312, data: 0.000) D_A: 0.103 G_A: 0.608 cycle_A: 0.790 idt_A: 0.224 D_B: 0.210 G_B: 0.455 cycle_B: 0.652 idt_B: 0.471 
(epoch: 177, iters: 600, time: 0.309, data: 0.000) D_A: 0.072 G_A: 0.591 cycle_A: 0.616 idt_A: 0.229 D_B: 0.188 G_B: 0.668 cycle_B: 0.592 idt_B: 0.175 
(epoch: 177, iters: 700, time: 0.310, data: 0.000) D_A: 0.063 G_A: 0.856 cycle_A: 0.766 idt_A: 0.181 D_B: 0.103 G_B: 0.693 cycle_B: 0.685 idt_B: 0.208 
(epoch: 177, iters: 800, time: 0.312, data: 0.000) D_A: 0.050 G_A: 0.838 cycle_A: 0.570 idt_A: 0.178 D_B: 0.140 G_B: 0.667 cycle_B: 0.591 idt_B: 0.143 
(epoch: 177, iters: 900, time: 0.312, data: 0.000) D_A: 0.045 G_A: 0.869 cycle_A: 0.556 idt_A: 0.144 D_B: 0.139 G_B: 0.583 cycle_B: 0.392 idt_B: 0.165 
(epoch: 177, iters: 1000, time: 0.310, data: 0.000) D_A: 0.072 G_A: 0.740 cycle_A: 0.568 idt_A: 0.169 D_B: 0.161 G_B: 0.644 cycle_B: 0.509 idt_B: 0.156 
(epoch: 177, iters: 1100, time: 0.312, data: 0.000) D_A: 0.180 G_A: 0.283 cycle_A: 0.672 idt_A: 0.333 D_B: 0.313 G_B: 0.673 cycle_B: 1.010 idt_B: 0.205 
(epoch: 177, iters: 1200, time: 0.310, data: 0.000) D_A: 0.043 G_A: 0.761 cycle_A: 0.774 idt_A: 0.214 D_B: 0.182 G_B: 0.338 cycle_B: 0.721 idt_B: 0.267 
(epoch: 177, iters: 1300, time: 0.312, data: 0.000) D_A: 0.061 G_A: 0.509 cycle_A: 0.709 idt_A: 0.133 D_B: 0.104 G_B: 0.358 cycle_B: 0.356 idt_B: 0.224 
(epoch: 177, iters: 1400, time: 0.311, data: 0.000) D_A: 0.085 G_A: 0.812 cycle_A: 0.623 idt_A: 0.175 D_B: 0.093 G_B: 0.406 cycle_B: 0.527 idt_B: 0.157 
(epoch: 177, iters: 1500, time: 0.311, data: 0.000) D_A: 0.208 G_A: 0.285 cycle_A: 0.686 idt_A: 0.141 D_B: 0.179 G_B: 0.436 cycle_B: 0.477 idt_B: 0.185 
(epoch: 177, iters: 1600, time: 0.310, data: 0.000) D_A: 0.108 G_A: 0.599 cycle_A: 0.764 idt_A: 0.221 D_B: 0.247 G_B: 0.519 cycle_B: 0.663 idt_B: 0.232 
(epoch: 177, iters: 1700, time: 0.309, data: 0.000) D_A: 0.104 G_A: 0.455 cycle_A: 0.421 idt_A: 0.137 D_B: 0.282 G_B: 0.311 cycle_B: 0.494 idt_B: 0.107 
(epoch: 177, iters: 1800, time: 0.311, data: 0.000) D_A: 0.118 G_A: 0.738 cycle_A: 0.750 idt_A: 0.208 D_B: 0.107 G_B: 0.448 cycle_B: 0.593 idt_B: 0.183 
(epoch: 177, iters: 1900, time: 0.310, data: 0.000) D_A: 0.124 G_A: 0.681 cycle_A: 0.907 idt_A: 0.176 D_B: 0.167 G_B: 0.357 cycle_B: 0.501 idt_B: 0.359 
(epoch: 177, iters: 2000, time: 0.312, data: 0.000) D_A: 0.043 G_A: 0.865 cycle_A: 0.629 idt_A: 0.181 D_B: 0.137 G_B: 0.373 cycle_B: 0.571 idt_B: 0.204 
(epoch: 178, iters: 100, time: 0.311, data: 0.000) D_A: 0.063 G_A: 0.646 cycle_A: 0.867 idt_A: 0.205 D_B: 0.152 G_B: 0.310 cycle_B: 0.605 idt_B: 0.270 
(epoch: 178, iters: 200, time: 0.311, data: 0.000) D_A: 0.057 G_A: 0.835 cycle_A: 0.778 idt_A: 0.128 D_B: 0.149 G_B: 0.398 cycle_B: 0.400 idt_B: 0.219 
(epoch: 178, iters: 300, time: 0.310, data: 0.000) D_A: 0.130 G_A: 0.699 cycle_A: 0.623 idt_A: 0.214 D_B: 0.150 G_B: 0.456 cycle_B: 0.685 idt_B: 0.168 
(epoch: 178, iters: 400, time: 0.311, data: 0.000) D_A: 0.043 G_A: 0.996 cycle_A: 0.800 idt_A: 0.156 D_B: 0.148 G_B: 0.511 cycle_B: 0.576 idt_B: 0.205 
(epoch: 178, iters: 500, time: 0.309, data: 0.000) D_A: 0.073 G_A: 0.502 cycle_A: 1.065 idt_A: 0.222 D_B: 0.242 G_B: 0.423 cycle_B: 0.660 idt_B: 0.332 
(epoch: 178, iters: 600, time: 0.311, data: 0.000) D_A: 0.064 G_A: 0.623 cycle_A: 0.517 idt_A: 0.162 D_B: 0.168 G_B: 0.372 cycle_B: 0.582 idt_B: 0.168 
(epoch: 178, iters: 700, time: 0.311, data: 0.000) D_A: 0.106 G_A: 1.042 cycle_A: 0.620 idt_A: 0.142 D_B: 0.137 G_B: 0.443 cycle_B: 0.482 idt_B: 0.155 
(epoch: 178, iters: 800, time: 0.310, data: 0.000) D_A: 0.060 G_A: 1.090 cycle_A: 0.678 idt_A: 0.174 D_B: 0.111 G_B: 0.278 cycle_B: 0.588 idt_B: 0.234 
(epoch: 178, iters: 900, time: 0.315, data: 0.000) D_A: 0.105 G_A: 0.434 cycle_A: 0.724 idt_A: 0.265 D_B: 0.102 G_B: 0.814 cycle_B: 0.875 idt_B: 0.254 
(epoch: 178, iters: 1000, time: 0.310, data: 0.000) D_A: 0.101 G_A: 0.538 cycle_A: 0.550 idt_A: 0.144 D_B: 0.094 G_B: 0.516 cycle_B: 0.436 idt_B: 0.121 
(epoch: 178, iters: 1100, time: 0.312, data: 0.000) D_A: 0.144 G_A: 0.969 cycle_A: 0.690 idt_A: 0.286 D_B: 0.189 G_B: 0.500 cycle_B: 0.845 idt_B: 0.213 
(epoch: 178, iters: 1200, time: 0.312, data: 0.000) D_A: 0.036 G_A: 0.535 cycle_A: 0.694 idt_A: 0.204 D_B: 0.161 G_B: 0.400 cycle_B: 0.528 idt_B: 0.193 
(epoch: 178, iters: 1300, time: 0.311, data: 0.000) D_A: 0.055 G_A: 1.121 cycle_A: 0.564 idt_A: 0.196 D_B: 0.074 G_B: 0.509 cycle_B: 0.690 idt_B: 0.180 
(epoch: 178, iters: 1400, time: 0.311, data: 0.000) D_A: 0.064 G_A: 0.714 cycle_A: 0.275 idt_A: 0.221 D_B: 0.097 G_B: 0.546 cycle_B: 0.623 idt_B: 0.104 
(epoch: 178, iters: 1500, time: 0.311, data: 0.000) D_A: 0.059 G_A: 0.828 cycle_A: 0.687 idt_A: 0.138 D_B: 0.201 G_B: 0.562 cycle_B: 0.396 idt_B: 0.227 
================ Training Loss (Sun Feb 13 08:10:29 2022) ================
**(epoch: 178, iters: 100, time: 0.328, data: 0.000) D_A: 0.451 G_A: 0.376 cycle_A: 3.520 idt_A: 2.585 D_B: 0.298 G_B: 0.491 cycle_B: 5.333 idt_B: 1.657** 
(epoch: 178, iters: 200, time: 0.301, data: 0.000) D_A: 0.245 G_A: 0.397 cycle_A: 2.881 idt_A: 0.912 D_B: 0.266 G_B: 0.281 cycle_B: 1.960 idt_B: 1.501 
(epoch: 178, iters: 300, time: 0.313, data: 0.000) D_A: 0.310 G_A: 0.522 cycle_A: 3.688 idt_A: 1.250 D_B: 0.295 G_B: 0.367 cycle_B: 2.599 idt_B: 1.805 
(epoch: 178, iters: 400, time: 0.315, data: 0.000) D_A: 0.384 G_A: 0.609 cycle_A: 1.985 idt_A: 1.352 D_B: 0.210 G_B: 0.281 cycle_B: 2.567 idt_B: 0.896 
(epoch: 178, iters: 500, time: 0.317, data: 0.000) D_A: 0.401 G_A: 0.304 cycle_A: 2.766 idt_A: 0.826 D_B: 0.406 G_B: 0.186 cycle_B: 1.873 idt_B: 1.138 
(epoch: 178, iters: 600, time: 0.297, data: 0.000) D_A: 0.252 G_A: 0.286 cycle_A: 2.035 idt_A: 1.149 D_B: 0.286 G_B: 0.238 cycle_B: 2.244 idt_B: 1.070 
(epoch: 178, iters: 700, time: 0.326, data: 0.000) D_A: 0.219 G_A: 0.228 cycle_A: 6.619 idt_A: 0.812 D_B: 0.144 G_B: 0.245 cycle_B: 1.724 idt_B: 2.757 
(epoch: 178, iters: 800, time: 0.320, data: 0.000) D_A: 0.199 G_A: 0.482 cycle_A: 2.067 idt_A: 1.852 D_B: 0.253 G_B: 0.205 cycle_B: 3.357 idt_B: 0.889 
(epoch: 178, iters: 900, time: 0.316, data: 0.000) D_A: 0.359 G_A: 0.197 cycle_A: 1.471 idt_A: 0.676 D_B: 0.333 G_B: 0.202 cycle_B: 1.403 idt_B: 0.699 
(epoch: 178, iters: 1000, time: 0.301, data: 0.000) D_A: 0.184 G_A: 0.209 cycle_A: 4.137 idt_A: 0.749 D_B: 0.378 G_B: 0.158 cycle_B: 1.711 idt_B: 1.994 
(epoch: 178, iters: 1100, time: 0.298, data: 0.000) D_A: 0.289 G_A: 0.313 cycle_A: 2.584 idt_A: 0.721 D_B: 0.201 G_B: 0.431 cycle_B: 1.532 idt_B: 1.272 
(epoch: 178, iters: 1200, time: 0.299, data: 0.000) D_A: 0.228 G_A: 0.430 cycle_A: 2.420 idt_A: 1.041 D_B: 0.220 G_B: 0.273 cycle_B: 1.911 idt_B: 1.074 
(epoch: 178, iters: 1300, time: 0.336, data: 0.000) D_A: 0.089 G_A: 0.488 cycle_A: 1.633 idt_A: 0.562 D_B: 0.155 G_B: 0.702 cycle_B: 1.250 idt_B: 1.042 
(epoch: 178, iters: 1400, time: 0.317, data: 0.000) D_A: 0.117 G_A: 0.425 cycle_A: 2.235 idt_A: 0.814 D_B: 0.165 G_B: 0.562 cycle_B: 1.710 idt_B: 0.791 
(epoch: 178, iters: 1500, time: 0.317, data: 0.000) D_A: 0.345 G_A: 0.537 cycle_A: 1.303 idt_A: 0.944 D_B: 0.073 G_B: 0.224 cycle_B: 1.893 idt_B: 0.723 
(epoch: 178, iters: 1600, time: 0.299, data: 0.000) D_A: 0.159 G_A: 0.472 cycle_A: 2.181 idt_A: 0.690 D_B: 0.213 G_B: 0.119 cycle_B: 1.835 idt_B: 0.949 
(epoch: 178, iters: 1700, time: 0.313, data: 0.000) D_A: 0.349 G_A: 0.179 cycle_A: 1.846 idt_A: 0.737 D_B: 0.184 G_B: 0.118 cycle_B: 1.614 idt_B: 0.889 
(epoch: 178, iters: 1800, time: 0.316, data: 0.000) D_A: 0.152 G_A: 0.142 cycle_A: 1.579 idt_A: 0.686 D_B: 0.295 G_B: 0.461 cycle_B: 2.521 idt_B: 0.745 
(epoch: 178, iters: 1900, time: 0.322, data: 0.000) D_A: 0.214 G_A: 0.241 cycle_A: 2.379 idt_A: 1.057 D_B: 0.204 G_B: 0.292 cycle_B: 2.144 idt_B: 1.020

**Comment by taesungp** [comment]
It is probably because the optimizer states of Adam are not saved in the checkpoints. Therefore, it's possible that the losses initially go a bit high for the first few iterations after resuming from checkpoints. Still, we found that it doesn't affect the overall training a lot because the losses stabilize fairly quickly. (disclaimer: we haven't done thorough study on this)

About the performance decrease between epoch 200 and 177, it could indeed be because of the unsaved optimizer states. Or it could be just because of the fluctuation during the course of training.

**Comment by nanditab35** [comment]
yes I have checked by putting breakpoints while saving and loading checkpoints, they have only model parameters for 4 models(netG_A, netG_B, netD_A, netD_B) but there is no optimizer parameters there for the 2 optimizers.

**Comment by junyanz** [comment]
You are right. The current code does not save the optimizer parameters. You probably have to modify the code.
